### ISSUE TITLE
League of Legends 5.14 | New Hud

### ISSUE THREAD

**Issue by Sayath** [tnajdek/lol-eyefinity-surround-fixhud#4]
Riot has released a new HUD for the LoL client. While the old HUD is still used in ranked games, all other game modes use the new version.

The new UI resides in menu/hud/renderui, which contains the following three files:
clarity_renderui.bin
clarity_renderui_flipped.bin
clarity_renderui_spectator.bin

In order to once again be able to center the UI, we have to decode the file format used. Specifically, how coordinates are stored in these three files.

**Comment by tnajdek** [comment]
I'll see what I can do.

**Comment by Sayath** [comment]
[New Hud with moved minimap frame](http://puu.sh/jcwbj/29be8e6446.png).

The binary file format is very similar to the old HUD. It follows roughly the same structure. Especially if you hex-diff clarity_renderui_spectator.bin, clarity_renderui.bin and clarity_renderui_flipped.bin together.

I'm still working out all the details. However, I've already deciphered the following structure:

``` c++
struct Header
{
    std::string type; // PROP - 50 52 4F 50
    uint32_t version; // 01 00 00 00 (version 1)
    uint32_t size; // 02 00 00 00 (2) | 2A 05 00 00 (1322)
    // trailing header unclear, e.g., 
    // A0 0E F0 3B A0 0E F0 3B
};

struct Rect
{
    float x_min;
    float y_min;
    float x_max;
    float y_max;
};

// pay attention to compiler padding
struct Interface_Element
{
    // initial section unclear, e.g., 
    // 89 00 00 00 84 68 C1 48 0C 00 5D 8B 22 6A 01 00 |
    // C0 00 00 00 53 81 F1 31 10 00 D1 95 E3 EE 01 01 |
    // 19 59 53 8A 01 01 B3 00 00 00 48 75 17 79 0F 00

    // 33 5B 6C C1 10 @ 978 < 1322 occurrences

    uint16_t texture_length;
    std::string texture;
    uint8_t ignored[5]; // DB BF EF 19 10
    uint16_t name_length;
    std::string name;
    uint8_t ignored2[5]; // E0 AC 91 8E 10
    uint16_t group_length;
    std::string group;

    // further understanding of the gap between group and rect required

    Rect rect;
    uint8_t ignored3[5]; // BB EF 24 2D 07
    uint32_t target_resolution_x; // 80 07 00 00 (1920) | 40 06 00 00 (1600)
    uint8_t ignored4[5]; // 56 C1 8A 34 07
    uint32_t target_resolution_y; // 38 04 00 00 (1080) | B0 04 00 00 (1200)
    // 19 59 53 8A 01 01 @ 647 < 1322 occurrences

    /*
    enabled;
    keepMaxScale;
    alwaysRespond;
    draggable;
    layer;
    anchor;
    alpha;
    uv;
    */
};

// for clarity_renderui_spectator | the actual ingame UI is found in clarity_renderui
Interface_Element minimapFrame;
minimapFrame.texture_length = 0x09; // 09 00
minimapFrame.texture = "Spectator";
minimapFrame.name_length = 0x0C; // 0C 00
minimapFrame.name = "MinimapFrame";
minimapFrame.group_length = 0x07; // 07 00
minimapFrame.group = "Minimap";
Rect rect;
rect.x_min = 1506.0; // 00 40 BC 44
rect.y_min = 769.0;  // 00 40 40 44
rect.x_max = 1920.0; // 00 00 F0 44
rect.y_max = 1080.0; // 00 00 87 44

minimapFrame.rect = rect;
```

In order to better understand the format, I suggest starting out with clarity_renderui_spectator, which has the lowest number of elements (only two, MinimapFrame and MinimapContents).

My goal is to decipher the format at least to an extent where we can automatically locate the rect and anchor attributes for each element. Then it should be possible to automatically centralize the HUD once again.

I've noticed that the rectangles have changed in size. Previously, the target resolution was 1024x768, now it's at least 1920x1080 or maybe even 1920x1200. Does this pose a problem for hud-fixer.py?

**Comment by tnajdek** [comment]
That's some really good work there, thanks for sharing! I'm currently off the grid but I'll look into this when possible.

**Comment by Sayath** [comment]
I've completed my Eyefinity fix:

Search for the following hashes:
- Anchor: 19 6B 7C 32 0B -> float anchor_x; float anchor_y;
- Rect: 76 BC F0 8F 0D -> float x_min; float y_min; float x_max; float y_max;

Furthermore, I've created a full specification of the new file structure [including all other attributes]. However, I'm missing some of the attribute names / their purpose is unclear, e.g., the boolean flags.

New HUD on Eyefinity:
- [before: spanning all monitors](http://puu.sh/jhNOU/2734dc28a5.jpg)
- [after: centered on primary screen](http://puu.sh/jhNNP/4e616c84db.jpg)

**Comment by BuzzBazz** [comment]
I'm having a small problem... How do I decipher the file "Clarity_RenderUI.bin" ?
When I open it in Notepad++ it becomes all gibberish as I have not been able to decipher it.
While I like the fact that you are making a program to fix the 3 or 5 monitor problem I would still like to be able to fix this on a manual basis if it where to break again.
Also, I have a question about the fix.
Will this fix the mouse edges as well? if it was possible to edit so that camera movement is restricted to only the center screen and not all three or five.

**Comment by Sayath** [comment]
Binary files (.bin) are not meant to be read by humans. If you ever have to edit them, you use a hex editor (notepad++ has a plugin you can install; however, it's pretty rudimentary).

Given the amount of individual UI element (greater 1200), a manual solution is tedious and error-prone. I'd advise against it. Use a programming language instead to automate the process.

To my knowledge, the mouse edge detection is hard-coded into the game client to equal the size of the window. Unless you are able to find a configuration file within the riot archive file structure, you'd have to modify the executable directly.

Reverse engineering assembler is a very complicated process. You'd have to translate the assembler to a higher level language (e.g., C++) and then find the right locations in the code that control mouse edge detection. There you'd inject your custom Eyefinity handling. Finally, you recompile your code into a standalone executable.

Personally, I'm using WASD for screen movement. Or if you prefer, you could also use ESDF. Remap the other keys appropriately. After a sufficient learning period, screen movement via keyboard is superior to mouse edge detection. Consequently, I have made no efforts to implement proper mouse edge detection for Eyefinity (and do not intend to in the foreseeable future).

**If you absolutely want to centralize the HUD manually:**

I'd advise reading [Centralizing the HUD](https://www.doppnet.com/hacking-league-of-legends-hud.html) first to get a basic understanding.

Download a great HEX editor (e.g., [010 Editor](http://www.sweetscape.com/010editor/)).

Then you just have to search for the following hashes (unique identifier sequence of 5 bytes) in the file:
- Anchor: 19 6B 7C 32 0B -> float anchor_x; float anchor_y;
- Rect: 76 BC F0 8F 0D -> float x_min; float y_min; float x_max; float y_max;

Find the information for each UI element, transform it according to the formula given in the article and then write it to the file.

Also the element anchored on the left side of the screen (target bar) has to be moved some additional units.

**Comment by BuzzBazz** [comment]
Yeah, I figured as much... Been a long time since I've looked at Binary files...
Hmm... didn't think that it would be that much a hassle to make a workaround for the mouse edge borders, but it makes sense that it would be something they would hard code.
Yeah, was thinking about setting up my keys so I can use the keyboard instead of the mouse for movement...
Read the whole post, that's why I'm here... As of right now none of the methods available are working and that's why I wanted to know a cruder fix for it if it was possible.
And it is, so no worries about nitty gritty as I like tinkering with stuff...
I got Wooxy to sort of work as the FAR Manager is "Offline" and in transfer to the Dragon Project, The only thing with Wooxy is that I could not get the skin I added to work, so I need to go back to the drawing board for that one.

**Comment by tnajdek** [comment]
Hey @BuzzBazz, I've got a clarity reader/writer mostly working here: https://github.com/tnajdek/lol-clarity

As soon as I find some time I'll integrate everything into a nice executable so it works as it did before.

**Comment by BuzzBazz** [comment]
Awesome :D Much obliged good sir :P
I just hope that Riot takes this seriously enough so that it makes it into the game code...
Or you get hired by them :P

**Comment by tnajdek** [comment]
This has been fixed in [version 0.4](https://github.com/tnajdek/lol-eyefinity-surround-fixhud/releases/tag/0.4)

**Comment by Pupix** [comment]
@Sayath The 5 bytes that you actually encounter are made up of `uint32 hash` + `uint8 identifier`.

The identifier specifies how the data block should be read. I have some of the identifiers mapped out, but there seem to be a lot more lurking around the files.

This is a barebone 010 Editor template for all .bin files: http://pastebin.com/raw.php?i=AwtEeqVi

All that's left is to match the hashes to their `string` counterpart and you could make a very good editor out of the available data inside the files.
